### PR TITLE
[9/N][Emit final] Emit final for session window aggregations

### PR DESCRIPTION
**Pull request by guozhangwang**
1. Add a new API for session windows to range query session window by end time (KIP related).
2. Augment session window aggregator with emit strategy.
3. Minor: consolidated some dup classes.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/state/SessionStore.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/SessionStore.java]: This is related to 1) in the description, and the first open question: is this public API worth to add? Note I added it into SessionStore not ReadOnlySessionStore, to not expose via IQv1, also I've only added this function for `Instant` param type as well.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/InMemorySessionStore.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/InMemorySessionStore.java:205]: This is a minor fix on the param names: the old ones are simply wrong and misleading.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimeOrderedSessionStore.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimeOrderedSessionStore.java]: This is the second open question: with the current prefixed (base, i.e. time-first) session key schema, this fetchAll would be effectively searching for `[earliestEnd, INF]` because of this logic: https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/state/internals/PrefixedSessionKeySchemas.java#L46

This is because we translate the range query without key inside `AbstractRocksDBTimeOrderedSegmentedBytesStore` by using the `lower/upperRange` instead of `lower/upperRangeFixedSize`): https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/state/internals/AbstractRocksDBTimeOrderedSegmentedBytesStore.java#L241-L242

I cannot remember why we need to do this. @lihaosky @mjsax do you remember why?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/SegmentedBytesStore.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/SegmentedBytesStore.java:94]: Minor javadoc improvement to remind developers.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/TimestampedTupleForwarder.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/TimestampedTupleForwarder.java:42]: This is 3) in the description: as we can use the base class, we then would not need the duplicated TupleFowarder any more.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java:208]: I realize that our SessionWindow and TimeWindow, and even SlidingWindow caused many code duplications (e.g. here) where we can just consolidate into the same class, with boolean flags indicating if the start/end are inclusive or exclusive, with that we can further reduce code duplication. Will file a JIRA for it.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java]: Will remove commented out code when removing WIP, ditto elsewhere.

**Review by guozhangwang — COMMENTED** [review]
@lihaosky @mjsax This PR is not mergable yet since I've not committed any new tests. But I'd like to solicit your feedbacks earlier, especially on two open question I made above (one about the public API, one about the inefficiency due to conservative upper/lowerRange with null key).

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java:104]: Should we not better pass in a `Time` object, so we can mock it using TTD?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java]: For what case could `windowedkey.key() == null` ? Is this even possible?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/AbstractKStreamTimeWindowAggregateProcessor.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/AbstractKStreamTimeWindowAggregateProcessor.java:79]: Thanks for all the cleanup -- it's somewhat distracting from the actual changes.

Can we (in the future) extract refactorings/cleanups into individual PRs to simplify reviewing?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimeOrderedSessionStore.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimeOrderedSessionStore.java]: Not sure -- I always need to think very hard to understand (not even sure if I succeed) the fetch logic and how we compute the bounds...

But same question as above: why do we need this new method instead of calling `findSessions(null, null, A, B)` -- I briefly dug into the code and it seems it would do the same thing?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/InMemorySessionStore.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/InMemorySessionStore.java]: Not sure if I fully understand why we add this new method instead of calling `findSession(null, null, A, B)` ?

The code to create the iterator is different, but I am also not sure why. Is it semantically actually the same? Calling `findSession(null, null, A, B)` would do:
```
registerNewIterator(null, // same
                    null, // same
                    latestSessionStartTime, // why does your code pass Long.MAX_VALUE,
// but because we use `tailMap` instead of `subMap` below, it seems to do the same thing overall?
                    endTimeMap.tailMap(earliestSessionEndTime, true).entrySet().iterator(),
                    true); // same
```

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/AbstractKStreamTimeWindowAggregateProcessor.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/AbstractKStreamTimeWindowAggregateProcessor.java:79]: Yes! It's my bad to mingle them together here.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/InMemorySessionStore.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/InMemorySessionStore.java]: Logically, the main reason is that for emit-final, we need a range query where the from/to are both `endTime`, i.e. you can see the parameters are `earliestSessionEndTime` and `latestSessionEndTime`.

Whereas for the existing functions, their semantics are based on `earliestSessionEndTime` but `latestSessionStartTime`. And that's also the reason for using `Long.MAX_VALUE` here.

On the physical implementation, the main difference is not in the in-memory session store, but the rocksDB session store. I will reply there separately.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimeOrderedSessionStore.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimeOrderedSessionStore.java]: The main reason is that for emit final I need a range API that based on `endTime` for both ends. And that's also why within its implementation I'd have to use `fetchAll` instead of `fetch` here.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/InMemorySessionStore.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/InMemorySessionStore.java]: Nit: can we call `subMap(earliestEndTime, true, latestEndTime, true)` which is the same thing but more "intuitive" as we always search for _inclusive_ bounds throughout the code (otherwise, this is the only place which has an exclusive upper bound).

File: streams/src/main/java/org/apache/kafka/streams/state/internals/InMemorySessionStore.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/InMemorySessionStore.java]: Ok. I read the code of `InMemorySessionStore` in detail and now understand what's going on. This LGTM.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/InMemorySessionStore.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/InMemorySessionStore.java]: nit: indention

File: streams/src/main/java/org/apache/kafka/streams/state/SessionStore.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/state/SessionStore.java]: I think there is no way around it? In the end, we allow users to plugin a custom session-store -- thus, if the use the new emit-final, why will need to implement this new method -- existing code with custom session-stores should not break, because existing code does neither implement but also not sure this new method.

If we don't make it public API, we would prevent users to pass in custom session-stores in combination with the new emit-final feature, what seems to be too restrictive?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimeOrderedSessionStore.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimeOrderedSessionStore.java]: If I read the code correctly, what `fetchAll()` does is correct: from my understanding, that fetchAll() is implement to find "overlapping sessions" given a lower and upper bound -- the lower bound must be smaller and session end and the upper bound must be smaller than session start to find an overlap. Because the upper bound compares to session start,, and we use the "base" we need to search the full "data/base part" of the store.

I guess the issue is, that you actually cannot use `fetchAll()` at all for our purpose here? Passing in `lastEndTime` does not work (does it) as it would be used to compare to session start-times, but we want to do a comparison to session end time. -- Thus, I think the right solution is, to actually also add the new `findSessions()` to the internal `SegmentedStore` and implement a proper iterator there?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java]: This behavior was meant to inherit from the deleted code: https://github.com/apache/kafka/pull/12204/files#diff-85c8c92d464af8eb3a60684bf929725f8fc5263353c38cacc20bee4cefe4fd9eL53, but after checking that logic I now realized it's not necessary anymore (the original PR has to do so since we cannot programmatically guarantee it's always not `null` but in this change we do not have that concern anymore).

Will remove.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/InMemorySessionStore.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/InMemorySessionStore.java]: Ack.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimeOrderedSessionStore.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimeOrderedSessionStore.java]: Sounds good. I would just have a special handling on both lower/upper bound as well as the hasNext function for this specific purpose.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java:104]: Makes sense. I will do this in a follow-up PR after merging this.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java]: nit: can this be simplified to `Math.max(0L, lastEmitWindowCloseTime)`? (Can also be address in follow up PR)

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java:265]: Should we use milliseconds or nanoseconds (I am always unsure)

File: streams/src/main/java/org/apache/kafka/streams/state/SessionStore.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/state/SessionStore.java]: Why did you pick `Instant` over `long` (wondering if `long` might be better as it's more an internal API)?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/PrefixedSessionKeySchemas.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/PrefixedSessionKeySchemas.java]: Can `key` ever be `null` here? (nit: add `{}` to block)

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimeOrderedSessionSegmentedBytesStore.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimeOrderedSessionSegmentedBytesStore.java]: nit: add `{}` to block

File: streams/src/main/java/org/apache/kafka/streams/state/internals/SessionKeySchema.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/SessionKeySchema.java:38]: Can `key` ever be `null` here?

**Review by mjsax — APPROVED** [review]
Overall LGTM.

A few nits. Also, there is some missing JavaDoc and stuff you did just put into comments but not remove yet. Also ok to cleanup in a follow up PR.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java]: Ack.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java:265]: Should be ms to be consistent with other metrics.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/PrefixedSessionKeySchemas.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/PrefixedSessionKeySchemas.java]: Yes it's possible since it's used to write lower/upper boundaries in which keys could be null.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/SessionKeySchema.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/SessionKeySchema.java:38]: Yes that's possible -- see above comment, for the lower/upper bound cases.

**Comment by guozhangwang** [comment]
Thanks @mjsax , I've addressed your comment, and also added the test coverage. While adding tests I noticed a bug in the code and fixed it (see my comment above).

Could you please take another look?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java:255]: This is one minor bug I detected in the latest commit.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimeOrderedSessionSegmentedBytesStore.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimeOrderedSessionSegmentedBytesStore.java:79]: This is the other minor bug I detected in the latest commit.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/SessionStoreBuilder.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/SessionStoreBuilder.java]: Here's the other change I made in order to work around the current tricky situation, since in-memory stores are always "time ordered" as well, we stripe the caching if the inner store is not persistent. cc @mjsax

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/SessionStoreBuilder.java]: If we strip the caching, this applies to eager emitting, right? -- So it would be a behavioral change? Do we want to piggy-back such a change into this KIP? Sounds "risky"?

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregateProcessorTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregateProcessorTest.java]: Might it be easier to change the internal config of the "emit interval" and set it to zero (instead of advancing time)?

**Comment by mjsax** [comment]
LGTM (assuming Jenkins passes).

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregateProcessorTest.java
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregateProcessorTest.java]: Ack!

File: streams/src/main/java/org/apache/kafka/streams/state/internals/SessionStoreBuilder.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/SessionStoreBuilder.java]: Sounds good, I will move this logic into the earlier stage during the topology building phase for the moment. Also cc @lihaosky who would do similar things for sliding windows.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/SessionWindowedKStreamImpl.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/SessionWindowedKStreamImpl.java:289]: cc @lihaosky this is what I did for session store during the topology building phase.

**Pull request merged**
